build.assets/tooling/lib/github/workflows.go

```diff
@@ -0,0 +1,180 @@
+	listOptions := github.ListWorkflowRunsOptions{
+		ListOptions: github.ListOptions{
+			PerPage: 100,
+		},
+		Branch:  ref,
+		Created: ">" + since.Format(time.RFC3339),
+	}
+
+	runIDs := make(map[int64]struct{})
+
+	for {
+		runs, resp, err := gh.client.Actions.ListWorkflowRunsByFileName(ctx, owner, repo, path, &listOptions)
+		if err != nil {
+			return nil, trace.Wrap(err, "Failed to fetch runs")
+		}
+
+		for _, r := range runs.WorkflowRuns {
+			runIDs[r.GetID()] = struct{}{}
+		}
+
+		if resp.NextPage == 0 {
+			break
+		}
+
+		listOptions.Page = resp.NextPage
+	}
+
+	return runIDs, nil
+}
+
+// TriggerDispatchEvent triggers a workflow_dispatch event in the target
+// repository and waits for a workflow to be started in response. Note that
+// this method requires that the GitHub and client clocks are roughly in sync.
+func (gh *ghClient) TriggerDispatchEvent(ctx context.Context, owner, repo, workflow, ref string, inputs map[string]interface{}) (*github.WorkflowRun, error) {
+	// There is no way that I know of to 100% accurately detect which workflow runs
+	// are created in response to a workflow_dispatch event. We can make a vey good
+	// guess, though, by looking at what workflow runs (with matching filename and
+	// source references) start immediately after we issue the event - so that's
+	// what we do here.
+
+	// Determine what workflows runs have already been created before we start, so
+	// we can exclude them when trying to detect a new run started in response to
+	// our dispatch event. Note that we pick a time slightly in the past to handle
+	// any clock skew.
+	baselineTime := time.Now().Add(-2 * time.Minute)
+	oldRuns, err := gh.ListWorkflowRuns(ctx, owner, repo, workflow, ref, baselineTime)
+	if err != nil {
+		return nil, trace.Wrap(err, "Failed to fetch task list")
+	}
+
+	log.Printf("Attempting to trigger %s/%s %s at ref %s\n", owner, repo, workflow, ref)
+	triggerArgs := github.CreateWorkflowDispatchEventRequest{
+		Ref:    ref,
+		Inputs: inputs,
+	}
+
+	// Issue the workflow_dispatch event.
+	_, err = gh.client.Actions.CreateWorkflowDispatchEventByFileName(ctx, owner, repo, workflow, triggerArgs)
+	if err != nil {
+		return nil, trace.Wrap(err, "Failed to issue request")
+	}
+
+	// Now we poll the GitHub API to see if any new Workflow Runs appear.We do this until
+	// the caller-supplied context expires, so be sure to set a timeout.
+	ticker := time.NewTicker(1 * time.Second)
+	defer ticker.Stop()
+
+	newRuns := make(map[int64]struct{})
+	for k, _ := range oldRuns {
+		newRuns[k] = struct{}{}
+	}
+
+	log.Printf("Waiting for new workflow run to start")
+
+	// Remember that the set of RunIDs includes completed runs as well as any
+	// in flight, so we don't have to account for expiring run IDs in our "old"
+	// set.
+	for setsAreEqual(newRuns, oldRuns) {
+		select {
+		case <-ticker.C:
+			newRuns, err = gh.ListWorkflowRuns(ctx, owner, repo, workflow, ref, baselineTime)
```

Comment:
clearly this is pretty racy, are we going to use this for actions that are run frequently? how bad is it if we pick the wrong run?

Reply:
It's definitely racy. The first use case is for things that will be run pretty infrequently, and with the git ref restriction covering them (e.g. merge builds and release tagging). But I'd not want to use it for jobs that were pinged all day.

Reply:
I'll investigate the tag method. It seems less terrible, and I'm happy to give up the ability to call an arbitrary workflow if it means we can just not worry about it guessing wrong.